### PR TITLE
Brunno

### PR DESCRIPTION
**Pull request by brunnoattorre**
New pull request with conflicts resolved + code review from @callicles

### CODE REVIEW

File: app/models/daos/ContributionDAO.scala [hunk 1]
```diff
@@ -96,4 +118,15 @@ class ContributionDAO @Inject() (neo: Neo4J){
...
   }
 
+  /**
+   * Parse multiple contributions and repos
+   * @param response response from neo4j
+   * @return map with each contribution from repo
+   */
+  def parseNeoContributions(response: WSResponse): Seq[(Repository,Contribution)] = {
+      ((Json.parse(response.body) \\ "row")).map{
+        case contribution => (contribution(0).as[Repository], contribution(1).as[Contribution])
```
Comment (callicles) [inline app/models/daos/ContributionDAO.scala]: Why is there a single case here ?

```scala
((Json.parse(response.body) \\ "row")).map({ contribution => (contribution(0).as[Repository], contribution(1).as[Contribution])}).seq
```

Comment (brunnoattorre) [inline app/models/daos/ContributionDAO.scala]: I didn't understand what you mean by single case. Each row will contain a tuple of (Repository,Contribution), so we are parsing each row and returning a Sequence of the tuples.

**Comment by brunnoattorre** [comment]
@callicles Fixed the issues you mentioned.

File: app/models/daos/UserDAO.scala [hunk 2]
```diff
@@ -129,45 +125,46 @@ class UserDAO @Inject() (neo: Neo4J) {
...
    * @return The parsed users.
    */
-  def parseNeoUsers(response: WSResponse): Seq[User] = {
-    var listUser = ArrayBuffer[User]()
-    ((((Json.parse(response.body) \ "results")(0) \ "data")(0) \ "row")(0) \\ "user").map(_.as[JsObject]).toList.foreach{
-        listUser +=  neo.parseSingleUser(_).get
-    }
-    listUser
-  }
+  def parseNeoUsers(response: WSResponse): Seq[User] = (Json.parse(response.body) \\ "user").map(neo.parseSingleUser(_).get)
```
Comment (callicles) [inline app/models/daos/UserDAO.scala]: Codacy Issue found: [Usage of get on optional type.](https://www.codacy.com/app/gitlinks/gitrank-web/file/938641865/issues/source?bid=2432310&fileBranchId=2846473#l128)

**Comment by brunnoattorre** [comment]
@callicles Please review :)

File: app/models/services/KarmaService.scala [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+  /**
+   * Calculate karma based on the user contributions
+   * @param user User to calculate
+   * @param contributions Map of Repos/Contributions
+   */
+
+  def calculateKarma(user: User, contributions: Seq[(Repository,Contribution)]): Int ={
+    contributions.map{
+      case (key,value) => (((value.addedLines.toFloat+value.removedLines)/(key.addedLines+key.removedLines)))*(key.score*key.score)
```
Comment (callicles) [inline app/models/services/KarmaService.scala]: Did not got that one last time :)

File: app/models/services/RepositoryService.scala [hunk 4]
```diff
@@ -114,7 +114,7 @@ class RepositoryService @Inject() (
    * @param repoName name of the repository to look for, "owner/repo"
    * @return A Sequence of contributors
    */
-  def findContributors(repoName: String): Future[Seq[User]] ={
+  def findContributors(repoName: String): Future[Seq[Option[User]]] ={
     repoDAO.find(repoName).map({
       case repo => return userDAO.findAllFromRepo(repo.get)
```
Comment (callicles) [inline app/models/services/RepositoryService.scala]: neither that one

File: app/models/services/RepositoryService.scala [hunk 5]
```diff
@@ -114,10 +114,8 @@ class RepositoryService @Inject() (
    * @param repoName name of the repository to look for, "owner/repo"
    * @return A Sequence of contributors
    */
-  def findContributors(repoName: String): Future[Seq[User]] ={
-    repoDAO.find(repoName).map({
-      case repo => return userDAO.findAllFromRepo(repo.get)
-    })
+  def findContributors(repoName: String): Future[Seq[Option[User]]] ={
+    repoDAO.find(repoName).map(repo => return userDAO.findAllFromRepo(repo.get))
```
Comment (callicles) [inline app/models/services/RepositoryService.scala]: Codacy Issue found: [Avoid using return](https://www.codacy.com/app/gitlinks/gitrank-web/file/939279771/issues/source?bid=2432310&fileBranchId=2846473#l118)

File: app/models/services/KarmaService.scala [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  }
+
+  /**
+   * Calculate karma based on the user contributions
+   * @param user User to calculate
+   * @param contributions Map of Repos/Contributions
+   */
+
+  def calculateKarma(user: User, contributions: Seq[(Repository,Contribution)]): Int ={
+    contributions.map(value => (((value._2.addedLines.toFloat+value._2.removedLines)/(value._1.addedLines+value._1.removedLines)))*(value._1.score*value._1.score)).sum.toInt
```
Comment (callicles) [inline app/models/services/KarmaService.scala]: Codacy Issue found: [Found unnamed result or parameter](https://www.codacy.com/app/gitlinks/gitrank-web/file/939279770/issues/source?bid=2432310&fileBranchId=2846473#l35)

**Pull request merged**